posts/python-in-production-part-1-of-5.md

```diff
@@ -0,0 +1,199 @@
+---
+authors:
+- Lucas Ward
+tags:
+- python
+- api
+- devops
+title: "Python in Production Part 1 of 5"
+image: 
+---
+
+Writing Python code is fast and easy.  It is thanks to this fact that Python as a language has gained immense popularity.  Putting Python code into production can also be fast and easy, if you follow a few guidelines.  In this series of blog posts, I will cover the how <em>and why</em> of several different aspects of writing <em>ready for production</em> python code. 
+
+Please note: this is not a Python programming best practices guide, but rather, a road map that any python project can follow to **ease the future push to production.** This series will teach you how to:
+* Avoid Dependency Issues
+* Write Portable Code
+* Package an Application
+* Integrate and Monitor an Application Using Systemd
+* Test and Build an Application
+
+For this series, I will be using a wonderful Python Web Framework called FastAPI.  If you are not using FastAPI but still want to follow along, fear not, these steps can be used for almost any type of application.  I cannot promise that these steps will make your application successful, but they will certainly make the launch, maintenance, and continued development much easier.  Most of these things I would consider as being <em>adjacent</em> to your Python application development.
+
+This series is made up of 5 Parts:
+1. Introduction and Setup, Using a Virtual Environment to Manage Dependencies.
+2. Creating Python Modules and Packages.
+3. Turning a Python Program into a Runnable Binary with Shiv.
+4. Running Python in Production with Systemd.
+5. Putting it all into a Build Pipeline with Jenkins.
+
+# Creating a Virtual Environment for Python
+
+The first step to creating a virtual environment requires us to think ahead, to the future.  What version of Python will this project ultimately run on? Will this project be a part of a legacy system running Python 2.7 or a cutting edge system running the latest version of Python 3?  Figuring this out ahead of time will make your life easier in the long run.  Once you have identified which version of Python you want to develop for, we need to make sure we have said version on our local machine.  If you are on a Mac, you likely already have Python installed. To check, simply run these two commands.
+
+To check which version of python you are running, type
+```bash
+python3 --version
+```
+
+For this series, we will be using Python version 3.10, although I will include some notes on how to get things working on earlier versions.  If you want to create a program to run on a different version of Pyton, you will need to make sure that you have that version of python installed, and that you create the environment using that version.
+
+To install Python 3.10 on a Mac, simply run the following command:
+```bash
+brew install python@3.10
```

Comment:
I'd consider enlightening people on `pyenv`  if you're up for the extra deets

Reply:
That would require my enlightenment as well. I will take a look!